internal/controller/challenges/apply_test.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2020 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package challenges
+
+import (
+	"strconv"
+	"sync"
+	"testing"
+
+	fuzz "github.com/google/gofuzz"
+	"github.com/stretchr/testify/assert"
+
+	cmacme "github.com/jetstack/cert-manager/pkg/apis/acme/v1"
+)
+
+func Test_serializeApply(t *testing.T) {
+	const (
+		expReg  = `^{"kind":"Challenge","apiVersion":"acme.cert-manager.io/v1","metadata":{.*},"spec":{.*},"status":{"processing":false,"presented":false}}$`
+		numJobs = 10000
+	)
+
+	var wg sync.WaitGroup
+	jobs := make(chan int)
+
+	wg.Add(numJobs)
+	for i := 0; i < 3; i++ {
+		go func() {
+			for j := range jobs {
+				t.Run("fuzz_"+strconv.Itoa(j), func(t *testing.T) {
+					var challenge cmacme.Challenge
+					fuzz.New().NilChance(0.5).Funcs(
+						func(challenge *cmacme.Challenge, c fuzz.Continue) {
+							if challenge.Spec.Solver.DNS01 != nil && challenge.Spec.Solver.DNS01.Webhook != nil {
+								challenge.Spec.Solver.DNS01.Webhook.Config = nil
+							}
+						},
+					).Fuzz(&challenge)
+
+					// Test regex with non-empty status.
+					challengeData, err := serializeApply(&challenge)
+					assert.NoError(t, err, "%+#v", challenge)
+					assert.Regexp(t, expReg, string(challengeData))
+
```

Comment:
Suggestion: add also roundtrip validation like you did in https://github.com/cert-manager/cert-manager/pull/4792/commits/5178ae2e49b3d80c3a86296a493ee0ea6b49d620#diff-352fae27eba791e80d7b2f03525372311950e0b1095f9612b44796bdae1a76c4R118-R120